pkg/app/launcher/launcher.go

```diff
@@ -0,0 +1,369 @@
+	}
+	if err := ensureDir(&launcher.conf.LocalPath); err != nil {
+		return nil, err
+	}
+
+	// Prepare networks.
+	skyN := appnet.NewSkywireNetworker(log.WithField("_", appnet.TypeSkynet), r)
+	if err := appnet.AddNetworker(appnet.TypeSkynet, skyN); err != nil {
+		return nil, err
+	}
+	dmsgN := appnet.NewDMSGNetworker(dmsgC)
+	if err := appnet.AddNetworker(appnet.TypeDmsg, dmsgN); err != nil {
+		return nil, err
+	}
+
+	// Kill old processes.
+	if err := launcher.killHangingProcesses(); err != nil {
+		return nil, err
+	}
+
+	// Prepare apps (autostart if necessary).
+	apps := make(map[string]AppConfig, len(conf.Apps))
+	for _, ac := range conf.Apps {
+		apps[ac.Name] = ac
+	}
+	launcher.apps = apps
+
+	return launcher, nil
+}
+
+// ResetConfig resets the launcher config.
+func (l *Launcher) ResetConfig(conf Config) {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	apps := make(map[string]AppConfig, len(conf.Apps))
+	for _, ac := range conf.Apps {
+		apps[ac.Name] = ac
+	}
+	l.apps = apps
+	l.conf = conf
+}
+
+// AutoStart auto-starts marked apps.
+func (l *Launcher) AutoStart(envMap map[string]func() []string) {
+	if envMap == nil {
+		envMap = make(map[string]func() []string)
+	}
+	log := l.log.WithField("func", "AutoStart")
+
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	for name, ac := range l.apps {
+		if !ac.AutoStart {
+			continue
+		}
+		var envs []string
+		if makeEnvs, ok := envMap[name]; ok {
+			envs = makeEnvs()
+		}
+		if err := l.startApp(name, ac.Args, envs); err != nil {
+			log.WithError(err).
+				WithField("app_name", name).
+				WithField("args", ac.Args).
+				WithField("envs", envs).
+				Warn("Failed to start app.")
+		}
+	}
+}
+
+// AppState returns a single app state of given name.
+func (l *Launcher) AppState(name string) (*AppState, bool) {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	ac, ok := l.apps[name]
+	if !ok {
+		return nil, false
+	}
+	state := &AppState{AppConfig: ac, Status: AppStatusStopped}
+	if _, ok := l.procM.ProcByName(ac.Name); ok {
+		state.Status = AppStatusRunning
+	}
+	return state, true
+}
+
+// AppStates returns list of AppStates for all registered apps.
+func (l *Launcher) AppStates() []*AppState {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	var states []*AppState
+	for _, app := range l.apps {
+		state := &AppState{AppConfig: app, Status: AppStatusStopped}
+		if _, ok := l.procM.ProcByName(app.Name); ok {
+			state.Status = AppStatusRunning
+		}
+		states = append(states, state)
+	}
+	return states
+}
+
+// StartApp starts cmd with given args and env.
+// If 'args' is nil, default args will be used.
+func (l *Launcher) StartApp(cmd string, args, envs []string) error {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	return l.startApp(cmd, args, envs)
+}
+
+func (l *Launcher) startApp(cmd string, args, envs []string) error {
+	log := l.log.WithField("func", "StartApp").WithField("cmd", cmd)
+
+	// Obtain associated app config.
+	ac, ok := l.apps[cmd]
+	if !ok {
+		return ErrAppNotFound
+	}
+	if args != nil {
```

Comment:
We should be checking `len(args) != 0` imo here. Just to exclude non-nil but still empty args

Reply:
@Darkren But what if people want to run an app with no args?